### PR TITLE
AP_SurfaceDistance: Handle bugs in abrupt reading changes

### PR DESCRIPTION
**Pull request by Georacer**
### Summary

This PR fixes the buggy behaviour of AP_SurfaceDistance (Surface Tracking) when flying with tethers and over vertical walls.
It introduces `SURFDSTD/U_` parameters to do so.

### Classification & Testing (check all that apply and add your own)

- [X] Checked by a human programmer
- [ ] Non-functional change
- [ ] No-binary change
- [ ] Infrastructure change (e.g. unit tests, helper scripts)
- [X] Automated test(s) verify changes (e.g. unit test, autotest)
- [X] Tested manually, description below (e.g. SITL)
- [ ] Tested on hardware
- [ ] Logs attached
- [X] Logs available on request

There are new autotests available. You can run them to observe the behaviour change before/after the fix.

#### Problem description

The surface tracking behaviour would be buggy in the following corner cases:

1. You fly over a building that is >2m high. The rangefinder readings usually do a step-change that is detected as a glitch. As a result the surface offset is reset and the vehicle does not climb to match the height of the building.
You then fly off the building, the rangefinder beam scanning its vertical wall. There is no step change and no glitch is detected. as a result, the vehicle descends even more and it's not in the initial altitude anymore.

Note: the 2m threshold is currently hardcoded. The glitch will cause a reset after 3 consecutive samples that exceed the threshold.

2. You fly with a tethered vehicle. Depending on how the tether flops around, abruptly getting into the FOV of the rangefinder and causing step changes in the measurements, it might:
- Cause the vehicle to temporarily ascend or descend, if the step change is below the 2m step threshold.
- Cause the vehicle to permanently ascend or descend, if the step change is above the 2m step threshold, through the same mechanism explained above.

I have had personal reports of both of these incidents from clients and partners.

You can see how these bugs play out in the 2nd commit, where autotests have been introduced.

#### Fix

The proposed solution is to add a glitch threshold parameter `GLDST` and a number of samples to trigger a reset parameter `GLSAM`. Setting either to 0 disables the respective feature.
Each is hosted in duplicate in the groups `SURFDSTU_` and `SURFDSTD_` groups, for the up-facing and down-facing rangefinder.

To address the building use-case, you increase `GLDST` to be higher than the highest building you plan to fly over, so that the step change will not be considered a glitch.

To address the tether use-case, you disable the `GLSAM` parameter, so that it never allows the reference altitude reset. It will ignore measurement step changes, but will not reset altitude because of them.

#### Testing

I have created an initial autotest that highlights the problem and intentionally fails. Then it is amended with the new parameters and passes.

#### Unknowns

I notice that not just `SURF[0].D` freezes and not updates when glitching, but also `RFND.Dist` (Arrow 2). I wasn't expecting that and I don't know why this happens.

Also, `XKF5.HAGL` seems to temporarily go bad and then recover (example in arrow 1). I don't know if this is expected or not, nor if this has other bad consequences.

image

---

I wanted to have shared parameters for both the up and down rangefinder, but I couldn't figure out how to do it without assigning them on top-level Copter and then passing their reference inside the `AP_SurfaceDistance` library, which seems wasteful.

### CODE REVIEW

File: libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp [hunk 1]
```diff
@@ -15,18 +15,48 @@
...
+
+    // @Param: GLDST
+    // @DisplayName: Glitch threshold for surface tracking
+    // @Description: When a rangefinder reading differs from the previous by more than this, it will be considered a glitch and will not be used. A value of zero disables this check.
+    // @Units: m
+    // @Range: 0 100
+    // @User: Advanced
+    AP_GROUPINFO("GLDST", 1, AP_SurfaceDistance, glitch_alt_m, RANGEFINDER_GLITCH_ALT_M_DEFAULT),
+
+    // @Param: GLDNS
```
Comment (tridge) [inline libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp]: GLDNS typo, should be GLSAM?

File: libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp [hunk 2]
```diff
@@ -15,18 +15,48 @@
...
+    // @DisplayName: Number of glitched samples before accepting the new reading
+    // @Description: When this many consecutive samples are considered a glitch, we give up and accept the new reading. A value of zero disables this check and a new value will never be accepted and the measurements will have to be below the glitch threshold in order to get used.
```
Comment (tridge) [inline libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp]: wording not very clear for the zero case

**Review by tridge — DISMISSED** [review]

File: Tools/autotest/arducopter.py [hunk 3]
```diff
@@ -5288,6 +5288,139 @@ def SurfaceTracking(self):
...
+    def SurfaceTrackingCornerCases(self):
+        """test the AP_SurfaceDistance library behaviour under real-life corner-cases"""
+
+        def restore_offset(steps):
+            """Gradually restore SIM_SONAR_OFFSET back to 0."""
+            initial_value = self.get_parameter("SIM_SONAR_OFFSET")
+            for i in range(steps):
+                self.set_parameter("SIM_SONAR_OFFSET", initial_value*(steps-1-i)/steps)
+
+        self.context_push()
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]: Don't need a context here.

File: Tools/autotest/arducopter.py [hunk 4]
```diff
@@ -5288,6 +5288,139 @@ def SurfaceTracking(self):
...
+        # The tether moves out of the FOV of the sensor.
+        self.set_parameter("SIM_SONAR_OFFSET", 0)
+        self.delay_sim_time(5) # Wait for the vehicle altitude to stabilize.
+        # We should still be at the same altitude.
+        self.assert_altitude(alt_initial, accuracy=0.5)
+        self.set_parameter("SURFDSTD_GLSAM", samples_default)  # Restore the default value.
+
+        self.do_RTL()
+        self.reboot_sitl()
+        self.context_pop()
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]:
```suggestion
```

File: libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp [hunk 5]
```diff
@@ -15,18 +15,48 @@
...
+// table of user settable parameters
+const AP_Param::GroupInfo AP_SurfaceDistance::var_info[] = {
+
+    // @Param: GLDST
+    // @DisplayName: Glitch threshold for surface tracking
+    // @Description: When a rangefinder reading differs from the previous by more than this, it will be considered a glitch and will not be used. A value of zero disables this check.
+    // @Units: m
+    // @Range: 0 100
+    // @User: Advanced
+    AP_GROUPINFO("GLDST", 1, AP_SurfaceDistance, glitch_alt_m, RANGEFINDER_GLITCH_ALT_M_DEFAULT),
```
Comment (peterbarker) [inline libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp]:
```suggestion
    AP_GROUPINFO("GLDST", 1, AP_SurfaceDistance, glitch_alt_m, AP_SURFACEDISTANCE_GLITCH_ALT_M_DEFAULT),
```

File: ArduCopter/Parameters.cpp [hunk 6]
```diff
@@ -1169,6 +1169,14 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
     // @Increment: 0.1
     AP_GROUPINFO("PILOT_TKO_ALT_M", 20, ParametersG2, pilot_takeoff_alt_m, PILOT_TKO_ALT_M_DEFAULT),
 
+    // @Group: SURFDSTD__
+    // @Path: ../libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp
+    AP_SUBGROUPPTR(surface_distance_down_ptr, "SURFDSTD_", 21, ParametersG2, AP_SurfaceDistance),
```
Comment (rmackay9) [inline ArduCopter/Parameters.cpp]: To keep the name short maybe the prefix should be "SURF_"

File: libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp (hunk 1 above)
Comment (rmackay9) [reply libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp]: maybe we should move the parameter to Copter and put it with the other SURFTRAK_ parameters?  So perhaps call it SURFTRAK_GLITCH or SURFTRAK_GLIT?

File: libraries/SITL/SITL.cpp [hunk 7]
```diff
@@ -658,6 +658,12 @@ const AP_Param::GroupInfo SIM::var_info3[] = {
 #ifdef SFML_JOYSTICK
     AP_SUBGROUPEXTENSION("",      63, SIM,  var_sfml_joystick),
 #endif // SFML_JOYSTICK
+    //
+    // @Param: SONAR_OFFSET
+    // @DisplayName: Sonar measurement offset.
+    // @Description: Sonar measurement offset, in meters. Can be used for error injection.
+    // @User: Advanced
+    AP_GROUPINFO("SONAR_OFFSET",     57, SIM,  sonar_offset, 0),
```
Comment (peterbarker) [inline libraries/SITL/SITL.cpp]: Make a SONAR sub-table

File: Tools/autotest/arducopter.py [hunk 8]
```diff
@@ -5288,6 +5288,139 @@ def SurfaceTracking(self):
...
+
+        self.progress("*** Fly with interference from an underslung tether.")
+        threshold = self.get_parameter("SURFDSTD_GLDST")
+        # Disable reference resets, we expect a lot of false glitches, due to cable interference.
+        self.set_parameter("SURFDSTD_GLSAM", 0)
+        # The tether moves in the FOV of the sensor.
+        self.set_parameter("SIM_SONAR_OFFSET", -(threshold+glitch_offset))
+        self.delay_sim_time(5) # Wait for the vehicle altitude to stabilize.
+        # The vehicle should reject the reading and keep still.
+        self.assert_altitude(alt_initial, accuracy=0.5)
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]:
```suggestion
        self.wait_altitude(alt_initial, accuracy=0.5, minimum_duration=5, timeout=10)
```

File: Tools/autotest/arducopter.py [hunk 9]
```diff
@@ -5288,6 +5288,139 @@ def SurfaceTracking(self):
...
+        # The tether moves out of the FOV of the sensor.
+        self.set_parameter("SIM_SONAR_OFFSET", 0)
+        self.delay_sim_time(5) # Wait for the vehicle altitude to stabilize.
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]:
```suggestion
```

File: Tools/autotest/arducopter.py [hunk 10]
```diff
@@ -5288,6 +5288,139 @@ def SurfaceTracking(self):
...
+        # We should still be at the same altitude.
+        self.assert_altitude(alt_initial, accuracy=0.5)
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]:
```suggestion
        self.wait_altitude(alt_initial-0.1, alt_initial+0.1, accuracy=0.5, miniumum_duration=x, timeout=y)
```

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp (hunk 1 above)
Comment (Georacer) [reply libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp]: I tried to use `AP_SUBGROUPVARPTR` in order to be able to have the parameter declarations inside `AP_SurfaceDistance`, but for the life of me I couldn't do it.

I fell back to your suggestion, Randy.

---

The GLDNS typo has been fixed.

File: libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp (hunk 2 above)
Comment (Georacer) [reply libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp]: Tried to clarify. Please read again.

File: ArduCopter/Parameters.cpp (hunk 6 above)
Comment (Georacer) [reply ArduCopter/Parameters.cpp]: I fell back to your suggestion to use "SURFTRAK".

File: Tools/autotest/arducopter.py (hunk 3 above)
Comment (Georacer) [reply Tools/autotest/arducopter.py]: I am a little concerned about the rangefinder params which I think need a reboot?

File: Tools/autotest/arducopter.py (hunk 4 above)
Comment (Georacer) [reply Tools/autotest/arducopter.py]: Same as my concern above. The previous test does have this pattern, most likely due to the rangefinder parameters.

File: libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp (hunk 5 above)
Comment (Georacer) [reply libraries/AP_SurfaceDistance/AP_SurfaceDistance.cpp]: Done.

File: libraries/SITL/SITL.cpp (hunk 7 above)
Comment (Georacer) [reply libraries/SITL/SITL.cpp]: Done.

File: Tools/autotest/arducopter.py (hunk 8 above)
Comment (Georacer) [reply Tools/autotest/arducopter.py]: Done.

File: Tools/autotest/arducopter.py (hunk 9 above)
Comment (Georacer) [reply Tools/autotest/arducopter.py]: Done.

File: Tools/autotest/arducopter.py (hunk 10 above)
Comment (Georacer) [reply Tools/autotest/arducopter.py]: Done.

File: ArduCopter/Parameters.cpp [hunk 11]
```diff
@@ -1169,6 +1169,25 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
     // @Increment: 0.1
     AP_GROUPINFO("PILOT_TKO_ALT_M", 20, ParametersG2, pilot_takeoff_alt_m, PILOT_TKO_ALT_M_DEFAULT),
 
+#if AP_RANGEFINDER_ENABLED
+    // @Param: SURFTRAK_GLDST
+    // @DisplayName: Glitch threshold for surface tracking
```
Comment (rmackay9) [inline ArduCopter/Parameters.cpp]: could we add Surface Tracking to the beginning of the display name?  So perhaps, "Surface Tracking Glitch threshold" and "Surface Tracking glitched sample count before target is reset"

File: ArduCopter/Parameters.h [hunk 12]
```diff
@@ -695,6 +695,8 @@ class ParametersG2 {
     AP_Int8                 failsafe_dr_enable;
     AP_Int16                failsafe_dr_timeout;
     AP_Float                surftrak_tc;
+    AP_Float                surftrak_glitch_alt;
+    AP_Int8                 surftrak_num_samples;
```
Comment (peterbarker) [inline ArduCopter/Parameters.h]:
```suggestion
    struct SurfTrakParameter {
        AP_Float                surftrak_glitch_alt;
        AP_Int8                 surftrak_num_samples;
    };
```

but declared in surface tracker library, it keeps a pointer to the structure passed in from the vehicle

(roughly speaking)

File: Tools/autotest/arducopter.py [hunk 13]
```diff
@@ -5383,6 +5288,132 @@ def SurfaceTracking(self):
...
+    def SurfaceTracking2(self):
+        """test the AP_SurfaceDistance library parameters"""
+
+        def restore_offset(steps):
+            """Gradually restore SIM_SONAR_OFFSET back to 0."""
+            initial_value = self.get_parameter("SIM_SONAR_OFFSET")
+            for i in range(steps):
+                self.set_parameter("SIM_SONAR_OFFSET", initial_value*(steps-1-i)/steps)
+
+        self.context_push()
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]:
```suggestion
```

File: Tools/autotest/arducopter.py [hunk 14]
```diff
@@ -5383,6 +5288,132 @@ def SurfaceTracking(self):
...
+        self.set_parameter("SIM_SONAR_OFFSET", -(threshold+glitch_offset))
+        self.delay_sim_time(5) # Wait for the vehicle altitude to stabilize.
+        self.assert_altitude(alt_initial, accuracy=0.5)  # The aircraft should reject the measurements and not climb.
+        restore_offset(10)
+        self.delay_sim_time(5) # Wait for the vehicle to stabilize again.
+        self.assert_altitude(alt_initial, accuracy=0.5)  # The aircraft should have maintained the same altitude reference.
+
+        self.do_RTL()
+        self.reboot_sitl()
+        self.context_pop()
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]:
```suggestion
```

**Review by rmackay9 — APPROVED** [review]

File: Tools/autotest/arducopter.py (hunk 3 above)
Comment (peterbarker) [reply Tools/autotest/arducopter.py]: Oh, you need the reboot.

But the harness sees you have done a reboot within your test and assumes you did that because you've set parameters that require parameters.

File: Tools/autotest/arducopter.py [hunk 15]
```diff
@@ -5383,6 +5288,132 @@ def SurfaceTracking(self):
...
+        self.set_parameter("SIM_SONAR_OFFSET", -(threshold-glitch_offset))
+        # This should result in increase of altitude.
+        target = alt_initial + (threshold - glitch_offset)
+        self.wait_altitude(target-0.5, target+0.5, minimum_duration=5, timeout=20)
+        # Fly off the building. The vertical wall is scanned by the beam, resulting in gradual reading increase.
+        restore_offset(10)
+        # We should be once again at the original altitude.
+        self.wait_altitude(alt_initial-0.5, alt_initial+0.5, minimum_duration=5, timeout=20)
+
+        self.progress("*** Fly with interference from an underslung tether.")
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]: You can use 
```suggestion
        self.start_subtest("Fly with interference from an underslung tether.")
```

**Review by peterbarker — COMMENTED** [review]
@MattKear FYI

**Comment by peterbarker** [comment]
Can be merged once the parameter passing done up as tridge described

**Comment by Georacer** [comment]
@peterbarker @tridge I'm much happier with the new parameter structure now. Thanks for the tip!

File: ArduCopter/Parameters.cpp [hunk 16]
```diff
@@ -1169,6 +1169,25 @@ const AP_Param::GroupInfo ParametersG2::var_info2[] = {
...
+    // @Description: When a rangefinder reading differs from the previous by more than this, it will be considered a glitch and will not be used. A value of zero disables this check.
+    // @Units: m
+    // @Range: 0 100
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("SURFTRAK_GLDST", 21, ParametersG2, surf_dist_parameters.glitch_alt, AP_SURFACEDISTANCE_GLITCH_ALT_M_DEFAULT),
+
+    // @Param: SURFTRAK_GLSAM
+    // @DisplayName: Number of glitched samples before accepting the new reading
```
Comment (rmackay9) [inline ArduCopter/Parameters.cpp]: could we change the DisplayName to "Surface Tracking Glitched Sample Count"

File: Tools/autotest/arducopter.py [hunk 17]
```diff
@@ -5469,6 +5288,129 @@ def SurfaceTracking(self):
...
+        self.set_analog_rangefinder_parameters()
+        self.reboot_sitl()
+
+        self.change_mode("LOITER")
+        self.wait_ready_to_arm()
+        self.takeoff(15, takeoff_throttle=1700, mode="LOITER")
+        alt_initial = self.get_altitude()
+        glitch_offset = 1
+
+        self.progress("*** Testing normal glitch procedure.")
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]:
```suggestion
        self.start_subtest("Testing normal glitch procedure.")
```

**Review by rmackay9 — APPROVED** [review]
Once the parameter display name is fixed feel free to merge!

**Review by peterbarker — APPROVED** [review]

File: Tools/autotest/arducopter.py (hunk 17 above)
Comment (Georacer) [reply Tools/autotest/arducopter.py]: Done.

File: ArduCopter/Parameters.cpp (hunk 16 above)
Comment (Georacer) [reply ArduCopter/Parameters.cpp]: Done.
